keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,641 @@
+  deprecated API. Except where we are explicitly documenting
+  Endpoints, no examples should use `kubectl get endpoints` or involve
+  creating an Endpoints object.
+
+- Write a blog post about the deprecation and the overall KEP plan,
+  and mention it in the "mid point comms" blog post prior to the
+  release.
+
+#### Deprecation - Stage 2
+
+Stage 2 can begin as soon as Stage 1 is complete, and does not need to
+be completed all at once.
+
+- Update all remaining internal code to use EndpointSlices rather than
+  Endpoints.
+
+- Update/reorganize e2e tests so that all tests that depend on the
+  Endpoints controller are in a single test suite in
+  `test/e2e/network/endpoints.go` and all tests that depend on the
+  EndpointSlice mirroring controller are in a single test suite in
+  `test/e2e/network/endpointslicemirroring.go`. (The latter may
+  already be true.)
+
+- Create a periodic e2e job that runs with the Endpoints and
+  EndpointSlice Mirroring controllers disabled, and with the
+  associated tests skipped, and confirm that it passes.
+
+#### Deprecation - Stage 3
+
+Stage 3 will not happen until the relevant SIGs feel that enough time
+has passed since the initial deprecation announcement.
+
+- The tests depending on the Endpoints and EndpointSlice Mirroring
+  controllers are demoted from conformance, and there is additional
+  communication (e.g. blog post) about this.
+
+- The documentation is updated to explain how to disable the Endpoints
+  and EndpointSlice Mirroring controllers, but warns that third-party
+  components may still depend on them.
+
+#### Deprecation - Stage 4
+
+Stage 4 will not happen until the relevant SIGs feel reasonably
+confident that the Kubernetes ecosystem has mostly moved away from
+Endpoints.
+
+- The documentation becomes more bullish on the idea of disabling the
+  controllers, implying that it's a reasonable default.
+
+### Upgrade / Downgrade Strategy
+
+The KEP does not propose any automatic change to behavior; behavior
+would only be changed when the administrator chose to disable the
+controllers, which could happen at any time.
+
+### Version Skew Strategy
+
+The only non-opt-in behavioral change is fixing the apiserver
+aggregation and proxying APIs to use EndpointSlices rather than
+Endpoints internally, which does not present any skew issues (since
+EndpointSlices have already existed for a long time at this point).
+
+(When the conformance criteria change to allow disabling the Endpoints
+controller, this would only apply to clusters that are _fully_ at the
+new version, to avoid skew issues. More specifically: it is not
+conforming to disable the Endpoints controller if you still have any
+apiservers that use Endpoints for apiserver aggregation and proxying.)
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+The KEP does not define a new feature, it just proposes that we
```

Comment:
Shouldn't we have a feature-gate for the changes we will do to stop using Endpoint in aggregation etc.?

Reply:
I was considering that a bugfix. The only way the behavior change could actually break someone is if they were writing out the Endpoints by hand rather than having them be generated from a Service _and_ they were explicitly setting the `skip-mirroring` annotation for some reason...

Reply:
OK - I don't feel strongly enough about it to fight for FG here.

But can you please extend the answer to this question to reflect your above comment?

Reply:
done